### PR TITLE
! Validation for the root uri in case it is redirected

### PR DESCRIPTION
**Pull request by mikegron**
https://coveord.atlassian.net/browse/CT-331

### CODE REVIEW

File: Abot/Crawler/WebCrawler.cs [hunk 1]
```diff
@@ -1024,6 +1031,53 @@ protected virtual void WaitMinimumRetryDelay(PageToCrawl pageToCrawl)
...
+        /// </remarks>
+        protected virtual Uri GetRedirectUri(CrawledPage crawledPage)
+        {
+            Uri locationUri;
+            if (_crawlContext.CrawlConfiguration.IsHttpRequestAutoRedirectsEnabled) {
+                // For auto redirects, look for the response uri.
+                locationUri = crawledPage.HttpWebResponse.ResponseUri;
+            } else {
+                // For manual redirects, we need to look for the location header.
+                var location = crawledPage.HttpWebResponse.Headers["Location"];
```
Comment (sdesilets) [inline Abot/Crawler/WebCrawler.cs]: Response.Headers[HttpResponseHeader.Location]

location can be null.

File: Abot/Crawler/WebCrawler.cs [hunk 2]
```diff
@@ -1024,6 +1031,53 @@ protected virtual void WaitMinimumRetryDelay(PageToCrawl pageToCrawl)
...
+                
+                if (!Uri.TryCreate(location, UriKind.Absolute, out locationUri))
```
Comment (sdesilets) [inline Abot/Crawler/WebCrawler.cs]: Will this throw is location is null?

File: Abot/Crawler/WebCrawler.cs (hunk 1 above)
Comment (mikegron) [inline Abot/Crawler/WebCrawler.cs]: Not if the request was redirected

File: Abot/Crawler/WebCrawler.cs (hunk 2 above)
Comment (mikegron) [inline Abot/Crawler/WebCrawler.cs]: it cannot, in theory!

File: Abot/Crawler/WebCrawler.cs [hunk 3]
```diff
@@ -1024,6 +1031,53 @@ protected virtual void WaitMinimumRetryDelay(PageToCrawl pageToCrawl)
...
+                {
+                    var site = crawledPage.Uri.Scheme + "://" + crawledPage.Uri.Host;
```
Comment (sdesilets) [inline Abot/Crawler/WebCrawler.cs]: The redirected page could be of a different scheme and could have a port. 
http://example.com:443/ => https://www.example.com/
http://example.com:8080/ => http://www,example.com:8080/

Wouldn't it be better to use locationUri.GetLeftPart(UriPartial.Authority) to get everything. if

File: Abot/Crawler/WebCrawler.cs [hunk 4]
```diff
@@ -1024,6 +1031,53 @@ protected virtual void WaitMinimumRetryDelay(PageToCrawl pageToCrawl)
...
+                    location = site + location;
+                    locationUri = new Uri(location);
```
Comment (sdesilets) [inline Abot/Crawler/WebCrawler.cs]: The redirected page could be of a different scheme and could have a port. 
http://example.com:443/ => https://www.example.com/
http://example.com:8080/ => http://www,example.com:8080/

Wouldn't it be better to use locationUri.GetLeftPart(UriPartial.Authority) to get this?

Comment (mikegron) [inline Abot/Crawler/WebCrawler.cs]: I guess since it does the two together:
location = crawledPage.Uri.GetLeftPart(UriPartial.Authority) + location;

**Pull request merged**